### PR TITLE
chore(build): add distribution files

### PR DESCRIPTION
**Pull request by vinifala**
Added distribution bundle.
https://github.com/atlassian/react-beautiful-dnd/issues/142

### CODE REVIEW

**Comment by alexreardon** [comment]
Does this naming convention satisfy the Clojurescript requirements? @Frozenlock?

**Comment by Frozenlock** [comment]
Thank you for doing this.

IIRC as long as one can easily generate a minified file for distribution we can package it with a minimum of manual operations.

I'll try to look at this in more details this evening and come back with an answer.

**Comment by alexreardon** [comment]
I think we also need to ensure that we are bundling react

**Comment by alexreardon** [comment]
Also, there is a lot of webpack stuff in there that we could avoid if we used [rollup](https://rollupjs.org/)

**Comment by Frozenlock** [comment]
After reviewing the process, the thing we need is the browserified file (like this [one](https://unpkg.com/react-datepicker@0.53.0/dist/)).

We provide our own packaged dependencies for the most popular libraries (React, react-motion, etc.).
This means that if they are in the _externals_ section of the webpack.config.js, we can avoid duplication.

However, I don't know enough about the vanilla JS ecosystem nor do I want to burden this project.
I would simply ask that you do distribution files as you normally would; we'll handle it from there.

**Comment by vinifala** [comment]
- [x] Swap out Webpack for Rollout
- [x] add dist to package.json#files
- [x] add dist to .gitignore
- [x] remove the dist directory from this commit

This is my first time using Rollup, I followed the implementation mentioned in https://github.com/atlassian/react-beautiful-dnd/pull/81

**Comment by alexreardon** [comment]
This is looking good. @treshugart does it look good to you?

@Frozenlock would this satisfy your requirements?

**Comment by vinifala** [comment]
Thank you @alexreardon. Please let me know if that suits your bill, @treshugart @Frozenlock :)

File: package.json [hunk 1]
```diff
@@ -15,16 +15,16 @@
     "url": "https://github.com/atlassian/react-beautiful-dnd/issues"
   },
   "main": "lib/index.js",
-  "files": [
-    "/lib"
-  ],
+  "files": ["/lib", "/dist"],
```
Comment (treshugart) [inline package.json]: Would be good to also add a `browser` field that maps to `dist/index.js` (it will be minified by bundlers, however unpkg won't serve the min file).

**Comment by treshugart** [comment]
LGTM but left a comment that we might think about adding now. Some things I'd suggest thinking about for the future:

- `dist/` is really `umd/` and can map to `browser` (I've commented about mapping the current `dist/` to that)
- `lib/` is really `node/` and can map to `main`
- `es/` that exports es5 (or lowest common denominator) + esm and can map to `module`
- Having an `esnext/` build for consuming the latest spec is something people are doing more and more. I've done this in Skate and map it to the `esnext` field.

**Comment by alexreardon** [comment]
@treshugart can we implement your suggestions in a follow up issue? It seems like the currently PR solves the immediate need

**Comment by treshugart** [comment]
> Some things I'd suggest thinking about for the future:

**Comment by treshugart** [comment]
Or if you mean the `browser` field, that's up to if you think people might be reaching into the dist manually and if you want to consider a folder rename a breaking change. I'm don't care what you do here.

File: rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+export default {
+  input: './lib/index.js',
+  output: {
+    file: `dist/index${args.min ? '.min' : ''}.js`,
+    format: 'umd',
+  },
+  name: 'ReactBeautifulDnd',
+  plugins: [
+    resolve({
+      browser: true,
```
Comment (alexreardon) [inline rollup.config.js]: why is this needed?

```js
// some package.json files have a `browser` field which
      // specifies alternative files to load for people bundling
      // for the browser. If that's you, use this option, otherwise
      // pkg.browser will be ignored
      browser: true,  // Default: false
```

Comment (alexreardon) [inline rollup.config.js]: it looks like this already defaults to true

```js
// use "main" field or index.js, even if it's not an ES6 module
      // (needs to be converted from CommonJS to ES6
      // – see https://github.com/rollup/rollup-plugin-commonjs
      main: true,  // Default: true
```

File: rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      main: true,
+    }),
+    commonjs({
```
Comment (alexreardon) [inline rollup.config.js]: not sure what this one is for. I think just listing react as an external is enough

Comment (vinifala) [reply rollup.config.js]: `rollup-plugin-commonjs` complains that React does not export `'Component', 'Children', 'createElement'` as named exports. This was the solution I've found.

Comment (vinifala) [reply rollup.config.js]: Listing React as an external dependency leads to
`(!) Missing global variable name
Use options.globals to specify browser global variable names corresponding to external modules
react (guessing '_react')`
I'm not sure if that's fine, or should I include `globals: { react: 'React' }` to `rollup.config.js`?

Comment (treshugart) [reply rollup.config.js]: Yeah, I think you're going to have to do that. I've found you can automate the dependencies part by doing something like:

```js
const pkg = require('./package.json');
const skip = {
  ...pkg.dependencies,
  ...pkg.devDependencies,
  ...pkg.optionalDependencies,
  ...pkg.peerDependencies
};
```

Jason Miller (of Preact fame) also shows this [here](https://gist.github.com/developit/41f088b6294e2591f53b).

**Comment by alexreardon** [comment]
This is looking good! I will give it a look over next week.

**Comment by vinifala** [comment]
Thanks @alexreardon

File: rollup.config.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import resolve from 'rollup-plugin-node-resolve';
+import commonjs from 'rollup-plugin-commonjs';
+import uglify from 'rollup-plugin-uglify';
+import babel from 'rollup-plugin-babel';
+import yargs from 'yargs';
+
+import pkg from './package.json';
+
+const args = yargs.argv;
+const external = Object.keys(pkg.dependencies || {}).concat(
```
Comment (alexreardon) [inline rollup.config.js]: Getting close! I think this should only include `react`. The other dependencies are internal and should not need to be provided by consumers.

Also, I am getting this warning:

```
(!) Missing global variable names
Use options.globals to specify browser global variable names corresponding to external modules
react (guessing '_react')
```

I would expect that this should be listed as 'React' in the `options.globals`

**Comment by alexreardon** [comment]
Thanks for preserving with this @vinifala

**Comment by vinifala** [comment]
Thanks @alexreardon, React added as only external dependency with proper global mapping :)

**Comment by vinifala** [comment]
`oci runtime error: exec failed: container_linux.go:265: starting container process caused "could not create session key: disk quota exceeded"`
Looks like we ran out of space.

**Comment by alexreardon** [comment]
Oh wow! I am not too worried about that at this stage

**Comment by alexreardon** [comment]
Well done @vinifala !! I had a look locally and it is looking good. I'll merge this pr and it will go out in the next release. Thanks for your ongoing efforts to get this across the line 🎉

**Comment by vinifala** [comment]
Thanks for your patience, @alexreardon. I'm looking forward to contribute further.
